File: Current version/java/PluginAdaptor/src/We3Graph/ServiceManager.java
package We3Graph;

import We3Graph.EventListeners.GraphLoadedEventListener;
import We3Graph.EventListeners.NewCommandsEventListener;
import We3Graph.Exceptions.AuthenicationException;
import org.json.JSONArray;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.Hashtable;
import java.util.concurrent.ConcurrentLinkedQueue;

/**
 * A class to handle graph interactions with the web service.
 */
class ServiceManager
{
    private String serviceURL_;
    private ConcurrentLinkedQueue<Command> outgoingCommandsQueue_;
    private int clientID_;
    private Thread syncIncommingThread_;
    private Thread syncOutgoingThread_;
    private AutoResetEvent sendEvent_;
    private int lastCommandIDReceived_;
    private boolean isLoading_;
    private int graphID_;
    private Integer userID_;
    private String whoToken_;
    private String graphAccessToken_;

    NewCommandsEventListener __newCommandsReceived;
    GraphLoadedEventListener __graphLoaded;

    ServiceManager(SystemManager systemManager, int graphID)
    {
        serviceURL_ = systemManager.GetServiceURL();
        userID_ = systemManager.__GetUserID();
        whoToken_ = systemManager.__GetWhoToken();
        graphID_ = graphID;
        lastCommandIDReceived_ = 0;
        isLoading_ = true;
        sendEvent_ = new AutoResetEvent(false);

        outgoingCommandsQueue_ = new ConcurrentLinkedQueue<>();

        try
        {
            getGraphAccessToken();
            getClientID();
        } catch (Exception e)
        {

        }
    }


    void __start(boolean receiveCommands)
    {
        if (receiveCommands)
        {
            syncIncommingThread_ = new Thread(new Runnable()
            {
                public void run()
                {
                    try
                    {
                        syncIncommingCommands();
                    } catch (Exception e)
                    {
                    }
                }
            });
            syncIncommingThread_.start();
        }
        syncOutgoingThread_ = new Thread(new Runnable()
        {
            public void run()
            {
                try
                {
                    syncOutgoingCommands();
                } catch (Exception e)
                {
                }
            }
        });
        syncOutgoingThread_.start();
    }

    private void getGraphAccessToken() throws AuthenicationException
    {
        String url = serviceURL_ + "graph-access-tokens";

        if (userID_ == null || whoToken_ == null)
            throw new AuthenicationException();

        Hashtable<String, String> urlParameters = new Hashtable<String, String>();

        JSONObject bodyParameters = new JSONObject();
        bodyParameters.put("UserID", userID_);
        bodyParameters.put("WhoToken", whoToken_);
        bodyParameters.put("GraphID", graphID_);
        bodyParameters.put("PermissionType", 1);


        JSONObject response = RESTHelper.SendTypicalRequest(url, urlParameters,
                "Post", bodyParameters);

        graphAccessToken_ = response.getString("GraphAccessToken");
    }

    private void getClientID() throws AuthenicationException
    {
        String url = serviceURL_ + "clients";

        if (userID_ == null || whoToken_ == null)
            throw new AuthenicationException();

        Hashtable<String, String> urlParameters = new Hashtable<String, String>();

        JSONObject bodyParameters = new JSONObject();
        bodyParameters.put("UserID", userID_);
        bodyParameters.put("WhoToken", whoToken_);
        bodyParameters.put("GraphID", graphID_);
        bodyParameters.put("ClientName", "Plugin");
        bodyParameters.put("GraphAccessToken", graphAccessToken_);

        JSONObject response = RESTHelper.SendTypicalRequest(url, urlParameters,
                "Post", bodyParameters);
        clientID_ = response.getInt("CreatedClientID");
    }

    int __getClientID()
    {
        return clientID_;
    }

    void __stop()
    {
        if (syncIncommingThread_ != null)
            syncIncommingThread_.stop();
        if (syncOutgoingThread_ != null)
            syncOutgoingThread_.stop();
    }

    private void syncIncommingCommands() throws AuthenicationException
    {
        while (true)
        {
            String url = serviceURL_ + "commands";

            if (userID_ == null || whoToken_ == null)
                throw new AuthenicationException();

            Hashtable<String, String> urlParameters = new Hashtable<String, String>();
            urlParameters.put("UserID", userID_.toString());
            urlParameters.put("WhoToken", whoToken_);
            urlParameters.put("GraphID", "" + graphID_);
            urlParameters.put("LastCommandID", "" + lastCommandIDReceived_);
            urlParameters.put("GraphAccessToken", graphAccessToken_);

            JSONObject bodyParameters = null;

            JSONArray response = RESTHelper.SendTypicalArrayRequest(
                    url, urlParameters, "Get", bodyParameters);
            processResponse(response);

            try
            {
                Thread.sleep(20);
            } catch (Exception e)
            {
            }
        }
    }

    private void syncOutgoingCommands() throws InterruptedException
    {
        while (true)
        {
            sendEvent_.waitOne(20);

            while (!outgoingCommandsQueue_.isEmpty())
            {
                Command command;
                command = outgoingCommandsQueue_.peek();
                sendCommand(command);
                outgoingCommandsQueue_.poll();
            }
        }

    }

    private void processResponse(JSONArray responseArray)
    {

        if (responseArray.length() == 0)
        {
            if (isLoading_)
            {
                isLoading_ = false;
                if (__graphLoaded != null)
                    __graphLoaded.OnGraphLoadedEvent();
            }
            return;
        }


        ArrayList<Command> commands = new ArrayList<Command>();

        for (int i = 0; i < responseArray.length(); i++)
        {
            JSONObject commandObject = responseArray.getJSONObject(i);

            String commandName = commandObject.getString("Name");
            int commandID = commandObject.getInt("ID");
            int clientID = commandObject.getInt("ClientID");
            ArrayList<String> parameters = new ArrayList<String>();
            for (int j = 1; j <= 5; j++)
            {
                String parameter = commandObject.getString("Param" + j);
                parameters.add(parameter);
            }

            Command command = new Command(commandName, parameters);
            command.__setID(commandID);
            command.__setClientID(clientID);

            commands.add(command);
            lastCommandIDReceived_ = command.__getID();
        }

        if (__newCommandsReceived != null)
            __newCommandsReceived.OnNewCommandEvent(commands);
    }

    void __runCommand(Command command)
    {
        outgoingCommandsQueue_.add(command);
        sendEvent_.set();
    }

    private boolean sendCommand(Command command)
    {
        try
        {
            ArrayList<String> parameters = command.GetParameters();

            String url = serviceURL_ + "commands";

            Hashtable urlParameters = new Hashtable<String, String>();

            JSONObject bodyParameters = new JSONObject();
            bodyParameters.put("UserID", userID_);
            bodyParameters.put("WhoToken", whoToken_);
            bodyParameters.put("GraphID", graphID_);
            bodyParameters.put("CommandName", command.GetName());
            bodyParameters.put("ClientID", clientID_);
            bodyParameters.put("GraphAccessToken", graphAccessToken_);
            bodyParameters.put("Param1", parameters.get(0));
            bodyParameters.put("Param2", parameters.get(1));
            bodyParameters.put("Param3", parameters.get(2));
            bodyParameters.put("Param4", parameters.get(3));
            bodyParameters.put("Param5", parameters.get(4));

            JSONObject response = RESTHelper.SendTypicalRequest(
                    url, urlParameters, "Post", bodyParameters);

            return true;
        } catch (Exception e)
        {
            return false;
        }

    }

    boolean __isAllCommandsSent()
    {
        if (outgoingCommandsQueue_.size() == 0)
            return true;
        else
            return false;
    }
}

File: Thesis version/c++/PluginAdaptor/ServiceManager.cc
#include "Servicemanager.h"
#include "Headers.h"

#include <unordered_map>
using namespace std;

namespace We3Graph
{
	ServiceManager::ServiceManager(SystemManager * systemManager, int graphID)
	{
		serviceURL_ = systemManager->GetServiceURL();
		userID_ = systemManager->__GetUserID();
		whoToken_ = systemManager->__GetWhoToken();
		graphID_ = graphID;
		lastCommandIDReceived_ = 0;
		isLoading_ = true;
		sendEvent_ = new AutoResetEvent(false);

		getGraphAccessToken();
		getClientID();
	}



	void ServiceManager::__start(bool receiveCommands)
	{
		if (receiveCommands)
		{
			syncIncommingThread_ = new thread(&ServiceManager::syncIncommingCommands, this);
		}
		syncOutgoingThread_ = new thread(&ServiceManager::syncOutgoingCommands, this);
	}

	void ServiceManager::getGraphAccessToken()
	{
		string url = serviceURL_ + "graph-access-tokens";

		if (userID_ == INT_MIN || whoToken_ == "")
			throw new AuthenicationException();

		unordered_map<string, string> urlParameters;
		json::value bodyParameters;

		bodyParameters[L"UserID"] = json::value::number(userID_);
		bodyParameters[L"WhoToken"] = json::value::string(
			Statics::StringToWide(whoToken_));
		bodyParameters[L"GraphID"] = json::value::number(graphID_);
		bodyParameters[L"PermissionType"] = json::value::number(1);

		auto response = RESTHelper::SendTypicalRequest(
			url, urlParameters, "Post", bodyParameters);
		graphAccessToken_ = Statics::WStringToNarrow(
			response[L"GraphAccessToken"].as_string());
	}

	void ServiceManager::getClientID()
	{
		string url = serviceURL_ + "clients";

		if (userID_ == INT_MIN || whoToken_ == "")
			throw new AuthenicationException();

		unordered_map<string, string> urlParameters;
		json::value bodyParameters;

		bodyParameters[L"UserID"] = json::value::number(userID_);
		bodyParameters[L"WhoToken"] = json::value::string(
			Statics::StringToWide(whoToken_));
		bodyParameters[L"GraphID"] = json::value::number(graphID_);
		bodyParameters[L"ClientName"] = json::value::string(L"Plugin");
		bodyParameters[L"GraphAccessToken"] = json::value::string(
			Statics::StringToWide(graphAccessToken_));

		auto response = RESTHelper::SendTypicalRequest(
			url, urlParameters, "Post", bodyParameters);
		clientID_ = response[L"CreatedClientID"].as_integer();
	}

	int ServiceManager::__getClientID()
	{
		return clientID_;
	}

	void ServiceManager::__stop()
	{
	}

	void ServiceManager::syncIncommingCommands()
	{
		while (true)
		{
			string url = serviceURL_ + "commands";

			if (userID_ == INT_MIN || whoToken_ == "")
				throw new AuthenicationException();

			unordered_map<string, string> urlParameters;

			urlParameters["UserID"] = to_string(userID_);
			urlParameters["WhoToken"] = whoToken_;
			urlParameters["GraphID"] = to_string(graphID_);
			urlParameters["LastCommandID"] = to_string(lastCommandIDReceived_);
			urlParameters["GraphAccessToken"] = graphAccessToken_;

			json::value bodyParameters;

			auto response = RESTHelper::SendTypicalRequest(
				url, urlParameters, "Get", bodyParameters);
			processResponse(response);

			this_thread::sleep_for(chrono::milliseconds(20));
		}
	}

	void ServiceManager::syncOutgoingCommands()
	{
		while (true)
		{
			sendEvent_->WaitOne(20);

			while (!outgoingCommandsQueue_.empty())
			{
				Command * command;
				outgoingCommandsQueueMutex_.lock();
				command = outgoingCommandsQueue_.front();
				outgoingCommandsQueue_.pop();
				outgoingCommandsQueueMutex_.unlock();
				sendCommand(command);
			}
		}

	}

	void ServiceManager::processResponse(json::value & response)
	{
		auto responseArray = response.as_array();

		if (responseArray.size() == 0)
		{
			if (isLoading_)
			{
				isLoading_ = false;
				if (__graphLoaded != NULL)
					__graphLoaded();
			}
			return;
		}


		auto commands = new vector<Command *>();

		for (auto item : responseArray)
		{
			string commandName = Statics::WStringToNarrow(item[L"Name"].as_string());
			int commandID = item[L"ID"].as_integer();
			int clientID = item[L"ClientID"].as_integer();
			auto parameters = new vector<string>();
			for (int i = 1; i <= 5; i++)
			{
				auto parameter = Statics::WStringToNarrow(
					item[L"Param" + to_wstring(i)].as_string());
				parameters->push_back(parameter);
			}

			auto command = new Command(commandName, parameters);
			command->__setID(commandID);
			command->__setClientID(clientID);

			commands->push_back(command);
			lastCommandIDReceived_ = command->__getID();
		}

		if (__newCommandsReceived != NULL)
			__newCommandsReceived(commands);
	}

	void ServiceManager::__runCommand(Command * command)
	{
		outgoingCommandsQueue_.push(command);
		sendEvent_->Set();
	}

	bool ServiceManager::sendCommand(Command * command)
	{
		try
		{
			auto parameters = command->GetParameters();

			string url = serviceURL_ + "commands";

			unordered_map<string, string> urlParameters;
			json::value bodyParameters;

			bodyParameters[L"UserID"] = json::value::number(userID_);
			bodyParameters[L"WhoToken"] = json::value::string(
				Statics::StringToWide(whoToken_));
			bodyParameters[L"GraphID"] = json::value::number(graphID_);
			bodyParameters[L"CommandName"] = json::value::string(
				Statics::StringToWide(command->GetName()));
			bodyParameters[L"ClientID"] = json::value::number(clientID_);
			bodyParameters[L"GraphAccessToken"] = json::value::string(
				Statics::StringToWide(graphAccessToken_));
			bodyParameters[L"Param1"] = json::value::string(
				Statics::StringToWide(parameters[0]));
			bodyParameters[L"Param2"] = json::value::string(
				Statics::StringToWide(parameters[1]));
			bodyParameters[L"Param3"] = json::value::string(
				Statics::StringToWide(parameters[2]));
			bodyParameters[L"Param4"] = json::value::string(
				Statics::StringToWide(parameters[3]));
			bodyParameters[L"Param5"] = json::value::string(
				Statics::StringToWide(parameters[4]));

			auto response = RESTHelper::SendTypicalRequest(
				url, urlParameters, "Post", bodyParameters);

			return true;
		}
		catch (exception e)
		{
			return false;
		}

		return false;
	}

	bool ServiceManager::__isAllCommandsSent()
	{
		if (outgoingCommandsQueue_.size() == 0)
			return true;
		else
			return false;
	}

}